Add filebeat_input to beat receiver agent monitoring integration test

leehinman opened the pull request:
## What does this PR do?

Adds last part of comparing process based monitoring vs beat receiver based monitoring

- Adds Windows as test case
- adds filebeat_input metrics to config
- adds filebeat_input metrics to integration test

## Why is it important?

We need to make sure that both implementations produce the same documents.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] I have read and understood the [pull request guidelines](https://github.com/elastic/elastic-agent/blob/main/CONTRIBUTING.md#pull-request-guidelines) of this project.
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## Disruptive User Impact

None.  Unit test and changes to producing configuration when experimental runtime is used.

## How to test this PR locally

```shell
mage integration:single TestClassicAndReceiverAgentMonitoring
```


## Related issues

- Closes #8188
- Closes elastic/ingest-dev#4876

## Questions to ask yourself

- How are we going to support this in production?
- How are we going to measure its adoption?
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

<!-- CI Cheatsheet
Trigger comments:
/test             (Or `buildkite test this|it`) Triggers unit test pipeline
/test extended    (Or `buildkite test extended`) Triggers integration test pipeline

PR labels:
skip-ci           Skips unit and integration tests
skip-it           Skips integration tests
-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/21824)
* Commit: f6d024aa1b4d0b99a53c516806eb0edb92b9ded0

### Failed CI Steps
* [Unit tests - Ubuntu 22.04](https://buildkite.com/elastic/elastic-agent/builds/21824#01973c24-e436-4aae-b341-ab4523748d54)
* [Unit tests - Ubuntu 22.04 with requirefips build tag](https://buildkite.com/elastic/elastic-agent/builds/21824#01973c24-e437-433a-883e-4a0c7dbb471a)
* [Unit tests - fips140=only Ubuntu 22.04](https://buildkite.com/elastic/elastic-agent/builds/21824#01973c24-e437-4337-ac8d-543ecc46f23e)
* [Unit tests - Ubuntu 22.04 ARM64](https://buildkite.com/elastic/elastic-agent/builds/21824#01973c24-e438-46df-8d9c-98af61a44658)
* [Unit tests - macOS 15 ARM](https://buildkite.com/elastic/elastic-agent/builds/21824#01973c24-e43b-49e5-9eb8-251e2c45ef54)
* [Unit tests - macOS 15 ARM](https://buildkite.com/elastic/elastic-agent/builds/21824#01973c35-fa7a-4e92-97e3-aa2564a92152)

cc @leehinman

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/21824","number":21824,"commit":"f6d024aa1b4d0b99a53c516806eb0edb92b9ded0"}],"number":21824}
buildkite-pr-comment-->

mauri870 commented on testing/integration/beat_receivers_test.go:
```diff
@@ -201,185 +238,225 @@ func TestClassicAndReceiverAgentMonitoring(t *testing.T) {
+			status, statusErr := receiverFixture.ExecStatus(ctx)
+			assert.NoError(collect, statusErr)
+			// agent should be healthy
+			assert.Equal(collect, int(cproto.State_HEALTHY), status.State)
```
You might want to apply these comments as the last argument of the assert or require functions. It will turn failure messages from something like "1 is not equal to 0" to "1 is not equal to 0 [...] agent status should be healthy" which is very helpful for debugging.

mauri870 commented on testing/integration/beat_receivers_test.go:
```diff
@@ -201,185 +238,225 @@ func TestClassicAndReceiverAgentMonitoring(t *testing.T) {
+			otelCollectorStatus := status.Collector
+			require.NotNil(collect, otelCollectorStatus)
+			assert.Equal(collect, int(cproto.CollectorComponentStatus_StatusOK), otelCollectorStatus.Status)
+			pipelineStatusMap := otelCollectorStatus.ComponentStatusMap
```
Can we use [`assertCollectorComponentsHealthy`](https://github.com/elastic/elastic-agent/blob/5648918/testing/integration/beat_receivers_test.go#L670) here?

mauri870 replied:
Ah nvm, I see you used it right below.

mauri870 commented on testing/integration/beat_receivers_test.go:
```diff
@@ -201,185 +238,225 @@ func TestClassicAndReceiverAgentMonitoring(t *testing.T) {
+				"@timestamp",
+				"agent.ephemeral_id",
+				"agent.id",
+				"agent.version",
```
If we are using the translation from the otel runtime then the versions should match, you could remove this field from the ignored list.

mauri870 replied:
FYI: This field can differ. The `agent.version` will always report the version of the Beats receiver. As of today, this will likely show 9.1.0. However, if you try to backport this to 9.0, the main Agent document will report 9.0, while the Beats receiver will still report 9.1.0.

mauri870 commented on testing/integration/beat_receivers_test.go:
```diff
@@ -201,185 +238,225 @@ func TestClassicAndReceiverAgentMonitoring(t *testing.T) {
+
+				// needs investigation
+				"event.agent_id_status",
+				"event.ingested",
```
event.ingested, event.duration is expected to change, so I don't think it needs investigation 🤔

khushijain21 commented on internal/pkg/agent/application/monitoring/v1_monitor.go:
```diff
@@ -701,7 +701,7 @@ func (b *BeatsMonitor) getHttpStreams(
 		}
 		// specifically for filebeat, we include input metrics
 		// disabled for filebeat receiver until https://github.com/elastic/beats/issues/43418 is resolved
```
we may want to remove this comment as well

swiatekm commented on testing/integration/otel_test.go:
```diff
@@ -1648,21 +1648,13 @@ func AssertMapsEqual(t *testing.T, m1, m2 mapstr.M, ignoredFields []string, msg
 	flatM1 := m1.Flatten()
 	flatM2 := m2.Flatten()
 	for _, f := range ignoredFields {
-		hasKeyM1, _ := flatM1.HasKey(f)
-		hasKeyM2, _ := flatM2.HasKey(f)
-
-		if !hasKeyM1 && !hasKeyM2 {
-			assert.Failf(t, msg, "ignored field %q does not exist in either map, please remove it from the ignored fields", f)
-		}
-
```
Why are we removing this? I thought it was here intentionally to ensure the ignored lists are trimmed whenever an element is not necessary,

leehinman replied:
I'm not in favor of coupling this test to beats implementation details.  If beats stops producing a field that we ignore, that isn't a failure for this test.  This test exists to help us make sure the documents are the same.  Both documents not having a field is the same.

That being said, I'll put it back, and we can reevaluate if it becomes a burden.

swiatekm reviewed (commented):
I don't get why changes to the existing test cases are necessary in this PR. Why does uncommenting the filebeat input test case not suffice?

mauri870 commented on testing/integration/beat_receivers_test.go:
```diff
@@ -201,185 +238,225 @@ func TestClassicAndReceiverAgentMonitoring(t *testing.T) {
+				// elastic_agent * fields are hardcoded in processor list for now which is why they differ
+				"elastic_agent.id",
+				"elastic_agent.snapshot",
+				"elastic_agent.version",
```
The `elastic_agent.version` field reports the version of Elastic Agent, which is the same in both tests. It should be safe to remove this field here so it gets compared properly.

leehinman commented:
> I don't get why changes to the existing test cases are necessary in this PR. Why does uncommenting the filebeat input test case not suffice?

2 things.

1) Because I found that the single query case wasn't sufficient, so I changed it to a slice

```
query           map[string]any
to
query           []map[string]any
```

2) Added Windows

elasticmachine commented:
Pinging @elastic/elastic-agent-control-plane (Team:Elastic-Agent-Control-Plane)

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/22061)
* Commit: c7865665758c4533bfd4f324f87a8f9c658ed78e

### Failed CI Steps
* [Unit tests - Ubuntu 22.04](https://buildkite.com/elastic/elastic-agent/builds/22061#01975602-61f7-4c6f-9563-994d64bb426c)
* [Unit tests - Ubuntu 22.04 with requirefips build tag](https://buildkite.com/elastic/elastic-agent/builds/22061#01975602-61f8-4b3d-b9ac-d5abe63f99a3)
* [Unit tests - fips140=only Ubuntu 22.04](https://buildkite.com/elastic/elastic-agent/builds/22061#01975602-61f9-46d6-b96c-fe0ebc7f5613)
* [Unit tests - Ubuntu 22.04 ARM64](https://buildkite.com/elastic/elastic-agent/builds/22061#01975602-61f9-4513-b4e7-c55efcda374a)
* [Unit tests - macOS 15 ARM](https://buildkite.com/elastic/elastic-agent/builds/22061#01975602-61fc-4895-950c-09a97637ffd2)
* [Unit tests - macOS 15 ARM](https://buildkite.com/elastic/elastic-agent/builds/22061#01975612-426a-4234-9564-d4a9c6370432)

### History
* :broken_heart: [Build #21824](https://buildkite.com/elastic/elastic-agent/builds/21824) failed f6d024aa1b4d0b99a53c516806eb0edb92b9ded0

cc @leehinman

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/22061","number":22061,"commit":"c7865665758c4533bfd4f324f87a8f9c658ed78e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/21824","number":21824,"commit":"f6d024aa1b4d0b99a53c516806eb0edb92b9ded0"}],"number":22061}
buildkite-pr-comment-->

leehinman commented:
@mauri870 @swiatekm @khushijain21 please take another look, should be clearer now.

swiatekm commented:
@leehinman you need to update the monitoring config generation unit tests.

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/22234)
* Commit: d403bf97e4ce4b4dd7a637cb61b9de4c68b0e555

### Failed CI Steps
* [fips:x86_64:sudo-false:fleet](https://buildkite.com/elastic/elastic-agent/builds/22234#0197606c-ff5b-4e7d-a522-6974bc070bf9)
* [fips:arm64:sudo-false:fleet](https://buildkite.com/elastic/elastic-agent/builds/22234#0197606c-ff5c-441c-bc03-9b9ca7848866)

### History
* :green_heart: [Build #22220](https://buildkite.com/elastic/elastic-agent/builds/22220) succeeded 0e2024f6d60ab18f704f5b42d7d0dd35d8b9f976
* :broken_heart: [Build #22210](https://buildkite.com/elastic/elastic-agent/builds/22210) failed b16e612d61a023e8c07f2ff9c754bf2c99006994
* :yellow_heart: [Build #22154](https://buildkite.com/elastic/elastic-agent/builds/22154) was flaky d38eed576eb762c004acb5d7720623e27cb33eff
* :broken_heart: [Build #22061](https://buildkite.com/elastic/elastic-agent/builds/22061) failed c7865665758c4533bfd4f324f87a8f9c658ed78e
* :broken_heart: [Build #21824](https://buildkite.com/elastic/elastic-agent/builds/21824) failed f6d024aa1b4d0b99a53c516806eb0edb92b9ded0

cc @leehinman

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/22234","number":22234,"commit":"d403bf97e4ce4b4dd7a637cb61b9de4c68b0e555"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/22220","number":22220,"commit":"0e2024f6d60ab18f704f5b42d7d0dd35d8b9f976"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/22210","number":22210,"commit":"b16e612d61a023e8c07f2ff9c754bf2c99006994"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/22154","number":22154,"commit":"d38eed576eb762c004acb5d7720623e27cb33eff"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/22061","number":22061,"commit":"c7865665758c4533bfd4f324f87a8f9c658ed78e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/21824","number":21824,"commit":"f6d024aa1b4d0b99a53c516806eb0edb92b9ded0"}],"number":22234}
buildkite-pr-comment-->

khushijain21 reviewed (approved):
Looks good to me!
